### DISCUSSION TITLE
Moving the choices example to a separate alpine.data

### DISCUSSION THREAD

**Discussion by aaronstezycki** [alpinejs/alpine#3282] [1. Help] [answered]
Hi there.

I've tried to move the choices example into a separate js file with an alpine.data configuration: [i.e.](https://alpinejs.dev/globals/alpine-data)

```js
import Choices from 'choices.js';

Alpine.data('select', (multiple, values, options, search = false, text = '' ) => ({
    init() {
        this.$nextTick(() => {
            let choices = new Choices(this.$refs.select, {
                classNames: {
                    containerOuter: 'c-input c-choices choices',
                },
                searchEnabled: search,
                itemSelectText: text,
            });

            let refreshChoices = () => {
                let selection = multiple ? values : [values]

                choices.clearStore()
                choices.setChoices(options.map(({value, label }) => ({
                    value,
                    label,
                    selected: selection.includes(value)
                })))
            }

            refreshChoices()

            this.$refs.select.addEventListener('change', () => {
                values = choices.getValue(true)
            })
            this.$watch('values', () => refreshChoices())
            this.$watch('options', () => refreshChoices())

        })
    }
}));
```

Then the markup:

```html
<div x-data="select(false, 1, [{ value: 1, label: 'Option 1' }, { value: 2, label: 'Option 2' }, { value: 3, label: 'Option 3' }] )">
    <label class="c-label">
        <span class="c-label__text">Label</span>
        <select x-ref="select"></select>
    </label>
</div>
```

Strangely, when loading the component it all works, however there are 2 console reference errors:
`module.esm.js:420 Uncaught ReferenceError: values is not defined`
and 
`Uncaught ReferenceError: options is not defined`

**Is there anything that is being done incorrectly here?**

[Choices example](https://alpinejs.dev/component/choices)

**Comment by ekwoka** [comment, marked as answer]
It's the `$watch` since you don't have a `values` or `options` prop on the component object.

Also, referencing them from the closure they won't be reactive.

You should simply put

```js
values,
options,
```

At the start of the component object and change your references to those to use `this.`

**Reply by aaronstezycki** [reply]
I've tried: 
```
this.multiple = multiple
this.values = values
this.options = options
```
and that works :) Thanks for your help. 

As an aside, if I wanted to retrieve the state of the multiple prop and set an attribute with it based on its value would I use it like so?...
`<select :multiple="select.multiple">`
I've tried this but no sauce.

**Reply by ekwoka** [reply]
Sure that works but why not just directly initialize it in the object like

```js
{
  multiple,
  values,
  options,
  init() {},
}
```


Etc.
Way cleaner.

**Reply by ekwoka** [reply]
Just do `:multiple="multiple"`

**Reply by aaronstezycki** [reply]
Thanks, I've gone with your approach, and yeah thats definitely cleaner. 
Ah, okay I did try that before but it failed, I'm guessing if I didn't declare the multiple prop as mentioned the attribute would fail also.

Thank you that really helped, and actually helped me fix some other components.
